pyomo/solvers/plugins/solvers/SAS.py

```diff
@@ -0,0 +1,701 @@
+        )
+
+        # Start a SAS session, submit the code and return the results``
+        with self._sas.SASsession() as sas:
+            # Find the version of 9.4 we are using
+            if sas.sasver.startswith("9.04.01M5"):
+                # In 9.4M5 we have to create an MPS data set from an MPS file first
+                # Earlier versions will not work because the MPS format in incompatible
+                res = sas.submit(
+                    """
+                                {warmstart}
+                                %MPS2SASD(MPSFILE="{mpsfile}", OUTDATA=mpsdata, MAXLEN=256, FORMAT=FREE);
+                                proc {proc} data=mpsdata {options} primalout=primalout dualout=dualout;
+                                {decomp}
+                                {decompmaster}
+                                {decompmasterip}
+                                {decompsubprob}
+                                {rootnode}
+                                run;
+                                """.format(
+                        warmstart=warmstart_str,
+                        proc=proc,
+                        mpsfile=self._problem_files[0],
+                        options=opt_str,
+                        decomp=decomp_str,
+                        decompmaster=decompmaster_str,
+                        decompmasterip=decompmasterip_str,
+                        decompsubprob=decompsubprob_str,
+                        rootnode=rootnode_str,
+                    ),
+                    results="TEXT",
+                )
+            else:
+                # Since 9.4M6+ optlp/optmilp can read mps files directly
+                res = sas.submit(
+                    """
+                                {warmstart}
+                                proc {proc} mpsfile=\"{mpsfile}\" {options} primalout=primalout dualout=dualout;
+                                {decomp}
+                                {decompmaster}
+                                {decompmasterip}
+                                {decompsubprob}
+                                {rootnode}
+                                run;
+                                """.format(
+                        warmstart=warmstart_str,
+                        proc=proc,
+                        mpsfile=self._problem_files[0],
+                        options=opt_str,
+                        decomp=decomp_str,
+                        decompmaster=decompmaster_str,
+                        decompmasterip=decompmasterip_str,
+                        decompsubprob=decompsubprob_str,
+                        rootnode=rootnode_str,
+                    ),
+                    results="TEXT",
+                )
+
+            # Store log and ODS output
+            self._log = res["LOG"]
+            self._lst = res["LST"]
+            # Print log if requested by the user
+            if self._tee:
```

Comment:
411 and 412 make more sense after 413's statement concludes; otherwise, they may be getting the full printout twice (once for tee, once for error).